Basic structure of a /workers/languages section (#13210)
cloudflare/cloudflare-docs: content/workers/languages/python/_index.md

---
pcx_content_type: navigation
title: Python
---

{{<heading-pill style="beta">}}Python{{</heading-pill>}}

Cloudflare Workers provides first-class support for Python, including:

> - [Built-in packages](https://pyodide.org/en/stable/usage/packages-in-pyodide.html)

Reviewer: Instead of pointing folks to Pyodide docs, should we have our own page with our own list? We will be starting out with a limited set of packages, so we may as well list them somewhere. These will include the packages listed in the Pyodide page but should include others too.

I think this is important to make the support feel more native.

Author: Yes I think this is great call out

> - Importing pure Python packages like [numpy](https://pypi.org/project/numpy/) from PyPI by adding them to `requirements.txt`

Reviewer: Might need a different example here, I may be wrong but I would expect `numpy` to very much not be a pure Python package (afaik it depends on C and Fortran libraries).

Author: @hoodmane - what's the best way to describe the scope of Python packages that will work? Had lifted this from our existing docs.

Reviewer 2: From Pyodide's docs:

> Pyodide supports installing following types of packages:
> * pure Python wheels from PyPI with [micropip](https://micropip.pyodide.org/en/v0.2.2/project/api.html#module-micropip).
> * pure Python and binary wasm32/emscripten wheels (also informally known as “Pyodide packages” or “packages built by Pyodide”) from the JsDelivr CDN and custom URLs.

But in workers, we will just have a list of packages. There's not really an obvious-to-end-user rule for what's in there.

Reviewer 3: Maybe a good story here is "we support this list of packages for now, but if you want a specific package you can request it through x"?

Author: ```suggestion
- Importing pure Python wheels from PyPI with micropip, by adding them to `requirements.txt`
```

Author: (will add link to request form further down)

- [FastAPI](https://fastapi.tiangolo.com/) compatibility, to build APIs in idiomatic Python
- [Langchain](https://pypi.org/project/langchain/) compatibility, to build LLM-powered apps easily

> - Bindings, Environment Variables, and Secrets

Author: Trying to bubble up here the important things that make this first-class and help highlight that this is beyond "it's just Wasm".

Reviewer: Maybe another point to include here is:
```suggestion
- Bindings, Environment Variables, and Secrets
- Workers API that you know and love, with a Pythonic feel
```

or something similar

> Python Workers are currently in open beta, and we'd love your feedback. Join the [#python-workers channel](https://discord.com/channels/595317990191398933/<TODO>) in the Cloudflare Discord community and let us know what you'd like to see next.

Author: @veroohh — can link to Discord channel once we have one

Reviewer: Please ping me when this is ready, want to make sure I have notifications enabled for that channel :)

## Get started

<!-- Run the following command in your terminal to create your first Python Worker: -->

<!-- TODO: cloudflare/workers-sdk/issues/5120 -->
<!-- ```bash

> npm create cloudflare@latest -- --template cloudflare/workers-sdk/templates/python

Author: refs EW-8201
refs https://github.com/cloudflare/workers-sdk/issues/5120

``` -->

```bash

> gh repo clone garrettgu10/python-workers-demos

Reviewer: Can this be changed to use `git` instead of `gh`? a lot more folks have `git` installed than `gh`.

> gh repo clone garrettgu10/python-workers-demos

Author: ```suggestion
git clone https://github.com/garrettgu10/python-workers-demos.git
```

cd hello
npx wrangler@latest dev
```

A Python Worker can be as simple as three lines of code:

```python
from js import Response

> def fetch(request):

Reviewer: Just an fyi: we'll rename this to `on_fetch`. CC EW-8183

Author: ```suggestion
def on_fetch(request):
```

    return Response.new("Hello World!")
```

> To run a Python Worker locally, you use [Wrangler](https://developers.cloudflare.com/workers/wrangler/), the CLI for Cloudflare Workers.

Author: ```suggestion
To run a Python Worker locally, you use [Wrangler](/workers/wrangler/), the CLI for Cloudflare Workers.
```

```bash
npx wrangler@latest dev
```

> 

Author: Below here is a TODO for @irvinebroque

Similar to Workers written in JavaScript or TypeScript, the main entry point for a Python worker is the
[`fetch` handler](/workers/runtime-apis/handlers/fetch). You'll notice that
the code looks very similar to a "Hello World" worker written in JavaScript.
In fact, if you have experience developing JavaScript workers, you should feel
right at home, since most of the functionality you're used to is still here.

## Using Modules

Python workers can also be split across multiple files. Let's create a new
Python file, called `hello.py`.

```python
def hello(name):
    return "Hello, " + name + "!"
```

Now, we can modify `entry.py` to make use of the new module.

```python
from hello import hello
from js import Response

> def fetch(request):

Author: ```suggestion
def on_fetch(request):
```

---
pcx_content_type: navigation

> title: Python

Author: ```suggestion
title: Python
weight: 3
```
